shared/src/main/scala/mlscript/NewParser.scala

```diff
@@ -800,24 +874,117 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
       case (tk, l0) :: _ =>
         err(msg"Unexpected ${tk.describe} in expression position" -> S(l0) :: Nil)
         consume
-        exprOrIf(prec)(et = et, fe = true, l = implicitly)
+        exprOrIf(prec)(et = et, fe = true, l = implicitly, qenv = qenv, quoted = quoted)
   }}
   
   private def errExpr =
     // Tup(Nil).withLoc(lastLoc) // TODO FIXME produce error term instead
     UnitLit(true).withLoc(lastLoc) // TODO FIXME produce error term instead
-  
-  final def exprCont(acc: Term, prec: Int, allowNewlines: Bool)(implicit et: ExpectThen, fe: FoundErr, l: Line): IfBody \/ Term = wrap(prec, s"`$acc`", allowNewlines) { l =>
+
+  private def quoteOp(lhs: Term, rhs: Term, op: Var) =
+    Quoted(App(op, PlainTup(Unquoted(lhs), Unquoted(rhs))))
+
+  private def declQenv(params: Term) = params match {
+    case Var(name) => Set[Str](name)
+    case Tup(params) => params.map {
+      case N -> Fld(_, Var(nme)) => nme
+      case S(Var(nme)) -> _ => nme
+      case _ => ???
+    }.toSet
+    case _ => Set.empty
+  }
+
+  private def arrowLam(genQuote: Bool, acc: Term, prec: Int, allowNewlines: Bool)(
+    implicit et: ExpectThen, fe: FoundErr, l: Line, qenv: Set[Str], quoted: IsQuoted
+  ): IfBody \/ Term = {
+    val newQenv = if (genQuote) declQenv(acc) else Set.empty
+    val rhs = expr(1)(fe, implicitly, qenv ++ newQenv, quoted)
+    if (!genQuote) exprCont(Lam(PlainTup(acc), rhs), prec, allowNewlines)
+    else {
+      val param = acc match {
+        case t: Tup => t
+        case _ => PlainTup(acc)
+      }
+      exprCont(Quoted(Lam(param, Unquoted(rhs))), prec, allowNewlines)
+    }
+  }
+
+  private def opCont(genQuote: Bool, opStr: Str, l0: Loc, acc: Term, prec: Int, allowNewlines: Bool)(
```

Comment:
Remove this extra function; use a `Ticked` syntax sugar term to represent ticked constructs.

Reply:
I found that the `Ticked` ast node didn't help a lot. So I just inserted unquotes after parsing. For operator applications, we need a helper function to handle expressions like `` `1 `* `2 `+ `3 `+ `4 ``. I still keep the `letbindings` function since we allow expressions like `` `let x = `0, y = `1 in x `+ y ``.